app/src/main/java/com/navapp/navigation/route/OptimizingRouteFactory.java

```diff
@@ -0,0 +1,61 @@
+package com.navapp.navigation.route;
+
+import com.google.maps.GeoApiContext;
+import com.google.maps.errors.ApiException;
+import com.navapp.navigation.destination.Destination;
+import com.navapp.navigation.destination.data.Location;
+import com.navapp.navigation.requests.RoutePlanner;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+public class OptimizingRouteFactory implements RouteFactory{
+    private final RoutePlanner routePlanner;
+
+    public OptimizingRouteFactory(GeoApiContext context) {
+        this(new RoutePlanner(context));
+    }
+
+    public OptimizingRouteFactory(RoutePlanner routePlanner) {
+        this.routePlanner = routePlanner;
+    }
+
+    @Override
+    public Route create(Location start, List<Destination> destinations, Location end) throws InterruptedException, ApiException, IOException {
+        List<String> destinationIds = extractIds(destinations);
+        List<String> orderedDestinationIds = routePlanner.orderedRouteIds(start.getId(), destinationIds, end.getId());
+        combineDestinationsWithNewIds(destinations, orderedDestinationIds, end);
+
+        return new Route(destinations);
+    }
+
+    private List<String> extractIds(List<Destination> destinations) {
+        List<String> destinationIds = new ArrayList<>();
+
+        for (Destination destination: destinations)
+            destinationIds.add(destination.getLocationId());
+
+        return destinationIds;
+    }
+
+    private void combineDestinationsWithNewIds(List<Destination> destinations, List<String> orderedIds, Location end) {
```

Comment:
Why not just create a new list instead of replacing stuff in this convoluted way?

Reply:
Technically it uses a bit less memory this way

Reply:
but yeah it's stupid